pkg/maps/ctmap/per_cluster_ctmap.go

```diff
@@ -102,379 +123,124 @@ type PerClusterCTMapVal struct {
-	var (
-		err error
-		im  *Map
+	gm.clusterIDs.Insert(clusterID)
+	return gm.foreach(
+		func(om *PerClusterCTMap) error { return om.createClusterCTMap(clusterID) },
 	)
-
-	ret := []*Map{}
-
-	gm.RLock()
-	defer gm.RUnlock()
-
-	defer func() {
-		if err != nil {
-			for _, im := range ret {
-				im.Unpin()
-				im.Close()
-			}
-		}
-	}()
-
-	if gm.ipv4 {
-		if im, err = gm.tcp4.getClusterMap(clusterID); err != nil {
-			return []*Map{}
-		} else {
-			ret = append(ret, im)
-		}
-		if im, err = gm.any4.getClusterMap(clusterID); err != nil {
-			return []*Map{}
-		} else {
-			ret = append(ret, im)
-		}
-	}
-
-	if gm.ipv6 {
-		if im, err = gm.tcp6.getClusterMap(clusterID); err != nil {
-			return []*Map{}
-		} else {
-			ret = append(ret, im)
-		}
-		if im, err = gm.any6.getClusterMap(clusterID); err != nil {
-			return []*Map{}
-		} else {
-			ret = append(ret, im)
-		}
-	}
-
-	return ret
 }
 
-func (gm *perClusterCTMaps) GetAllClusterCTMaps() ([]*Map, error) {
-	var err error
-	ret := []*Map{}
-
-	gm.RLock()
-	defer gm.RUnlock()
-
-	defer func() {
-		if err != nil {
-			for _, im := range ret {
-				im.Close()
-			}
-		}
-	}()
-
-	if gm.ipv4 {
-		if ims, err := gm.tcp4.getAllClusterMaps(); err != nil {
-			return nil, err
-		} else {
-			ret = append(ret, ims...)
-		}
-		if ims, err := gm.any4.getAllClusterMaps(); err != nil {
-			return nil, err
-		} else {
-			ret = append(ret, ims...)
-		}
-	}
-
-	if gm.ipv6 {
-		if ims, err := gm.tcp6.getAllClusterMaps(); err != nil {
-			return nil, err
-		} else {
-			ret = append(ret, ims...)
-		}
-		if ims, err := gm.any6.getAllClusterMaps(); err != nil {
-			return nil, err
-		} else {
-			ret = append(ret, ims...)
-		}
-	}
-
-	return ret, nil
-}
-
-func (gm *perClusterCTMaps) Cleanup() {
-	if gm.ipv4 {
-		gm.tcp4.cleanup()
-		gm.any4.cleanup()
-		gm.tcp4 = nil
-		gm.any4 = nil
-	}
-	if gm.ipv6 {
-		gm.tcp6.cleanup()
-		gm.any6.cleanup()
-		gm.tcp6 = nil
-		gm.any6 = nil
-	}
-}
-
-// Init a "dummy" global per-cluster CT maps
-func InitDummyPerClusterCTMaps(ipv4, ipv6 bool) {
-	PerClusterCTMaps = newDummyPerClusterCTMaps(ipv4, ipv6)
-}
-
-func newDummyPerClusterCTMaps(ipv4, ipv6 bool) *dummyPerClusterCTMaps {
-	gm := &dummyPerClusterCTMaps{
-		ipv4: ipv4,
-		ipv6: ipv6,
-	}
-
-	if ipv4 {
-		gm.tcp4 = make(map[uint32]struct{})
-		gm.any4 = make(map[uint32]struct{})
-	}
-
-	if ipv6 {
-		gm.tcp6 = make(map[uint32]struct{})
-		gm.any6 = make(map[uint32]struct{})
-	}
-
-	return gm
-}
-
-func (gm *dummyPerClusterCTMaps) UpdateClusterCTMaps(clusterID uint32) error {
+func (gm *perClusterCTMaps) DeleteClusterCTMaps(clusterID uint32) error {
 	if err := cmtypes.ValidateClusterID(clusterID); err != nil {
 		return err
 	}
 
 	gm.Lock()
 	defer gm.Unlock()
 
-	if gm.ipv4 {
-		gm.tcp4[clusterID] = struct{}{}
-		gm.any4[clusterID] = struct{}{}
-	}
-
-	if gm.ipv6 {
-		gm.tcp6[clusterID] = struct{}{}
-		gm.any6[clusterID] = struct{}{}
-	}
-
-	return nil
+	gm.clusterIDs.Delete(clusterID)
```

Comment:
Should we rollback this Delete when the below operation failed?

Reply:
Similar to the above comment, in case we attempted to delete the given maps, it likely makes sense not to return them anymore, regardless of whether the operation failed or not.